File: python/src/palimpsest/contracts/validation.py
from __future__ import annotations

import argparse
import base64
import json
from dataclasses import dataclass
from pathlib import Path
from typing import Any

from jsonschema import ValidationError

from .archive import canonical_archive_bytes
from .canonical_json import (
    ContractInputError,
    canonical_json_bytes,
    child_pointer,
    parse_json_strict,
)
from .digest import sha256_hex
from .gate_report_core import predeclaration_digest
from .schemas import CONTRACT_SCHEMAS, contract_validator


@dataclass(frozen=True)
class ValidationVerdict:
    accepted: bool
    reason: str | None
    pointer: str | None
    value: Any
    canonical_base64: str | None
    sha256: str | None


def _reject(reason: str, pointer: str) -> ValidationVerdict:
    return ValidationVerdict(False, reason, pointer, None, None, None)


def _pointer(error: ValidationError) -> str:
    pointer = "".join(child_pointer("", part) for part in error.absolute_path)
    if error.validator == "required":
        missing = error.message.split("'")[1]
        return child_pointer(pointer, missing)
    if error.validator in {"additionalProperties", "unevaluatedProperties"}:
        unknown = error.message.split("'")[1]
        return child_pointer(pointer, unknown)
    return pointer


def _reason(error: ValidationError, pointer: str) -> str:
    if pointer == "/schemaVersion":
        return "schema_version"
    if error.validator in {"additionalProperties", "unevaluatedProperties"}:
        return "unknown_field"
    if error.validator == "required":
        return "required"
    if error.validator == "type":
        return "type"
    if error.validator in {"enum", "const"}:
        return "enum"
    if error.validator in {"minimum", "maximum", "minItems", "maxItems", "minLength", "maxLength"}:
        return "range"
    return "format"


def validate_value(contract_id: str, value: Any) -> ValidationVerdict:
    if not isinstance(value, dict) or value.get("schemaVersion") != 1:
        return _reject("schema_version", "/schemaVersion")

    if contract_id == "canonical-archive":
        try:
            canonical_archive_bytes(value)
        except ContractInputError as error:
            return _reject(error.reason, error.pointer)

    errors = []
    for error in contract_validator(contract_id, value).iter_errors(value):
        pointer = _pointer(error)
        errors.append((pointer, _reason(error, pointer)))
    if errors:
        pointer, reason = sorted(errors)[0]
        return _reject(reason, pointer)

    if contract_id == "gate-report" and value.get("predeclarationDigest") != predeclaration_digest(
        value
    ):
        return _reject("digest", "/predeclarationDigest")

    canonical = canonical_json_bytes(value)
    return ValidationVerdict(
        True,
        None,
        None,
        value,
        base64.b64encode(canonical).decode("ascii"),
        sha256_hex(canonical),
    )


def validate_fixture(contract_id: str, raw: str) -> ValidationVerdict:
    if contract_id not in CONTRACT_SCHEMAS:
        return _reject("enum", "/contractId")
    try:
        value = parse_json_strict(raw)
    except ContractInputError as error:
        return _reject(error.reason, error.pointer)
    return validate_value(contract_id, value)


def build_fixture_verdicts(fixtures_root: Path) -> list[dict[str, Any]]:
    manifest = json.loads((fixtures_root / "manifest.json").read_text(encoding="utf-8"))
    verdicts = []
    for fixture in manifest["fixtures"]:
        raw = (fixtures_root / fixture["inputPath"]).read_text(encoding="utf-8")
        verdict = validate_fixture(fixture["contractId"], raw)
        archive_base64 = None
        archive_sha256 = None
        if verdict.accepted and fixture["contractId"] == "canonical-archive":
            archive = canonical_archive_bytes(verdict.value)
            archive_base64 = base64.b64encode(archive).decode("ascii")
            archive_sha256 = sha256_hex(archive)
        verdicts.append(
            {
                "accepted": verdict.accepted,
                "archiveBase64": archive_base64,
                "archiveSha256": archive_sha256,
                "canonicalBase64": verdict.canonical_base64,
                "fixtureId": fixture["fixtureId"],
                "pointer": verdict.pointer,
                "reason": verdict.reason,
                "sha256": verdict.sha256,
            }
        )
    return verdicts


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--fixture-verdicts", action="store_true")
    parser.add_argument("--values-file", type=Path)
    args = parser.parse_args()
    if args.values_file is not None:
        cases = json.loads(args.values_file.read_text(encoding="utf-8"))
        print(
            json.dumps(
                [
                    {
                        "accepted": verdict.accepted,
                        "pointer": verdict.pointer,
                        "reason": verdict.reason,
                        "sha256": verdict.sha256,
                    }
                    for case in cases
                    for verdict in [validate_value(case["contractId"], case["value"])]
                ],
                separators=(",", ":"),
            )
        )
        return
    if not args.fixture_verdicts:
        parser.error("--fixture-verdicts or --values-file is required")
    fixtures_root = Path(__file__).resolve().parents[4] / "packages" / "contracts" / "fixtures"
    print(
        json.dumps(
            build_fixture_verdicts(fixtures_root),
            ensure_ascii=False,
            separators=(",", ":"),
        )
    )


if __name__ == "__main__":
    main()

File: packages/contracts/src/validation.ts
import type { ErrorObject } from "ajv";

import { canonicalArchiveBytes } from "./archive.js";
import {
  ContractInputError,
  canonicalJsonBytes,
  childPointer,
  parseJsonStrict,
  type ContractReason,
} from "./canonical-json.js";
import { sha256Hex } from "./digest.js";
import { predeclarationDigest } from "./gate-report-core.js";
import { getContractValidator, isContractId, type ContractId } from "./schema-registry.js";

export interface AcceptedVerdict {
  accepted: true;
  canonicalBase64: string;
  pointer: null;
  reason: null;
  sha256: string;
  value: unknown;
}

export interface RejectedVerdict {
  accepted: false;
  canonicalBase64: null;
  pointer: string;
  reason: ContractReason;
  sha256: null;
  value: null;
}

export type ValidationVerdict = AcceptedVerdict | RejectedVerdict;

function reasonForError(error: ErrorObject): ContractReason {
  if (
    error.instancePath === "/schemaVersion" ||
    (error.keyword === "required" && error.params.missingProperty === "schemaVersion")
  ) {
    return "schema_version";
  }
  if (error.keyword === "additionalProperties" || error.keyword === "unevaluatedProperties") {
    return "unknown_field";
  }
  if (error.keyword === "required") {
    return "required";
  }
  if (error.keyword === "type") {
    return "type";
  }
  if (error.keyword === "enum" || error.keyword === "const") {
    return "enum";
  }
  if (
    error.keyword === "minimum" ||
    error.keyword === "maximum" ||
    error.keyword === "minItems" ||
    error.keyword === "maxItems" ||
    error.keyword === "minLength" ||
    error.keyword === "maxLength"
  ) {
    return "range";
  }
  return "format";
}

function pointerForError(error: ErrorObject): string {
  if (error.keyword === "required") {
    return childPointer(error.instancePath, String(error.params.missingProperty));
  }
  if (error.keyword === "additionalProperties" || error.keyword === "unevaluatedProperties") {
    return childPointer(
      error.instancePath,
      String(error.params.additionalProperty ?? error.params.unevaluatedProperty),
    );
  }
  return error.instancePath;
}

function reject(reason: ContractReason, pointer: string): RejectedVerdict {
  return {
    accepted: false,
    canonicalBase64: null,
    pointer,
    reason,
    sha256: null,
    value: null,
  };
}

function validateSchemaVersion(value: unknown): RejectedVerdict | undefined {
  if (
    value === null ||
    typeof value !== "object" ||
    !("schemaVersion" in value) ||
    value.schemaVersion !== 1
  ) {
    return reject("schema_version", "/schemaVersion");
  }
}

export function validateValue(contractId: ContractId, value: unknown): ValidationVerdict {
  const versionRejection = validateSchemaVersion(value);
  if (versionRejection) {
    return versionRejection;
  }

  if (contractId === "canonical-archive") {
    try {
      canonicalArchiveBytes(value);
    } catch (error) {
      if (error instanceof ContractInputError) {
        return reject(error.reason, error.pointer);
      }
      throw error;
    }
  }

  const validator = getContractValidator(contractId, value);
  if (!validator(value)) {
    const normalized = (validator.errors ?? [])
      .map((error) => ({
        pointer: pointerForError(error),
        reason: reasonForError(error),
      }))
      .sort(
        (left, right) =>
          left.pointer.localeCompare(right.pointer) || left.reason.localeCompare(right.reason),
      );
    const first = normalized[0];
    if (!first) {
      return reject("format", "");
    }
    return reject(first.reason, first.pointer);
  }

  if (
    contractId === "gate-report" &&
    value !== null &&
    typeof value === "object" &&
    "predeclarationDigest" in value &&
    value.predeclarationDigest !== predeclarationDigest(value as Record<string, unknown>)
  ) {
    return reject("digest", "/predeclarationDigest");
  }

  const bytes = canonicalJsonBytes(value);
  return {
    accepted: true,
    canonicalBase64: bytes.toString("base64"),
    pointer: null,
    reason: null,
    sha256: sha256Hex(bytes),
    value,
  };
}

export function validateFixture(contractId: string, raw: string): ValidationVerdict {
  if (!isContractId(contractId)) {
    return reject("enum", "/contractId");
  }
  let value;
  try {
    value = parseJsonStrict(raw);
  } catch (error) {
    if (error instanceof ContractInputError) {
      return reject(error.reason, error.pointer);
    }
    throw error;
  }
  return validateValue(contractId, value);
}

interface FixtureCase {
  contractId: string;
  expected: {
    accepted: boolean;
  };
  fixtureId: string;
  inputPath: string;
}

interface FixtureManifest {
  fixtures: FixtureCase[];
}

export async function buildFixtureVerdicts(fixturesRoot: URL): Promise<Record<string, unknown>[]> {
  const { readFile } = await import("node:fs/promises");
  const manifest = JSON.parse(
    await readFile(new URL("manifest.json", fixturesRoot), "utf8"),
  ) as FixtureManifest;
  const verdicts: Record<string, unknown>[] = [];
  for (const fixture of manifest.fixtures) {
    const raw = await readFile(new URL(fixture.inputPath, fixturesRoot), "utf8");
    const verdict = validateFixture(fixture.contractId, raw);
    let archiveBase64: string | null = null;
    let archiveSha256: string | null = null;
    if (verdict.accepted && fixture.contractId === "canonical-archive") {
      const archive = canonicalArchiveBytes(verdict.value);
      archiveBase64 = archive.toString("base64");
      archiveSha256 = sha256Hex(archive);
    }
    verdicts.push({
      accepted: verdict.accepted,
      archiveBase64,
      archiveSha256,
      canonicalBase64: verdict.accepted ? verdict.canonicalBase64 : null,
      fixtureId: fixture.fixtureId,
      pointer: verdict.pointer,
      reason: verdict.reason,
      sha256: verdict.accepted ? verdict.sha256 : null,
    });
  }
  return verdicts;
}

if (
  import.meta.url === `file://${process.argv[1]}` &&
  process.argv.includes("--fixture-verdicts")
) {
  const fixturesRoot = new URL("../fixtures/", import.meta.url);
  process.stdout.write(`${JSON.stringify(await buildFixtureVerdicts(fixturesRoot))}\n`);
}
